### ISSUE TITLE
upeptide_mapper fails with uparam write_unfiltered_results

### ISSUE THREAD

**Issue by davidyang10** [ursgal/ursgal#103]
While attempting a cascaded search, engine upeptide_mapper is failing under the ucontroller function uc.execute_misc_engine. The failure occurs while preparing the unode run for engine upeptide_mapper, and only fails when used for files obtained from parameter write_unfiltered_results. Failure occurs at the set_ios function within the ucontroller, where it appears that md5 files for the filtered result files being input into the upeptide_mapper engine are not being generated/missing. 
 
I have previously used uc.execute_misc_engine with files obtained from parameter write ‘unfiltered results’ with no issue however I used the engine merge_csv instead of upeptide_mapper. The error is probably due to differences in the execute_unode and merge_csv functions within the execute_misc_engine function in the uncontroller. 

The error message is shown here:
```
Traceback (most recent call last):
File "do_it_all_folder_wide_cascade_search_GlycoPA2.py", line 401, in <module>
  target_decoy_database=sys.argv[3],
File "do_it_all_folder_wide_cascade_search_GlycoPA2.py", line 323, in main
  scan_exclusion_list=list(spectra_with_PSM)
File "do_it_all_folder_wide_cascade_search_GlycoPA2.py", line 90, in workflow
  engine='upeptide_mapper',
File "/home/davyang/scripts/ursgal/ucontroller.py", line 965, in execute_misc_engine
  merge_duplicates = merge_duplicates,
File "/home/davyang/scripts/ursgal/ucontroller.py", line 2929, in execute_unode
  force  = force
File "/home/davyang/scripts/ursgal/ucontroller.py", line 1026, in prepare_unode_run
  userdefined_output_fname = output_file
File "/home/davyang/scripts/ursgal/ucontroller.py", line 1314, in set_ios
  self.io['input']['finfo']['md5'] = self.io['input']['o_finfo']['md5']
KeyError: 'md5'

```
The part of the script I used that is failing shown here:

```
if search_engine == 'msfragger_20170103':
    if level != '0':
        uc.params['csv_filter_rules']=[
            ['Spectrum ID', 'contains_element_of_list', scan_exclusion_list],
        ]
        uc.params['write_unfiltered_results'] = True
        
        search_result_intermediate = uc.execute_misc_engine(
            input_file=converted_result,
            engine='filter_csv',
        )
        converted_result = search_result_intermediate.replace('accepted', 'rejected')
        uc.params['csv_filter_rules'] = []

mapped_results = uc.execute_misc_engine(
    input_file=converted_result,
    engine='upeptide_mapper',
)

```

**Comment by fu** [comment]
Hi David,

could you please share input files and script to reproduce the error?

Thanks

.c

**Comment by fu** [comment]
Hi David,

@StSchulze bounced the files to me (again) - so sorry for the late reply. I looked into the issue and it was quite dazzling. One thing that I understood is that the json and the csv do not belong to each other, which might cause the funny behaviour. So the file and json ends with 

`_20170103_rejected.csv`

yet the json itself says
 
`_20170103_accepted.csv`

Then, as you pointed out the md5 is missing and that should be the case as long the processing node did not finish, in which case "run_status" would still be on "scheduled", which it is. I guess the node that should have created "_accepted" was never finished. So I suspect that the next node to filter for rejected elements was not given the right input, maybe?

Hope that helps,

Cheers

.c

**Comment by StSchulze** [comment]
Just as a clarification (without me having actually the files from the whole pipeline to test):
The *_accepted.csv and *_rejected.csv are both generated by the same node (filter_csv), so there would not be a json generated for the *_rejected.csv, right (only an input json once a new engine starts to use it, bot not an output json from the filter_csv node)?

And, to my understanding, the upeptide_mapper should not need the output json from the previous engine, correct?

**Comment by fu** [comment]
Yes and no. So the demo file comes with a json and that is lacking a md5 because the node that produced the corresponding file was not finished correctly. Now, as I said there is some kinda mismatch because the output file has a different file root than the 2nd element in the json list. The jsons are structured like this

`[`
`  { # 1. file information from the source .. i.e. what was used as input for the node },`
`  { # 2. file information from the target ... i.e. the output of the node which created that file},`
`  { # 3. params and history }`
`]`
`

The 2nd block should be identical to the actual file it is associated with 
file_x.csv
file_x.csv.u.json
yet it reads
file_rejected.csv
but the 2nd block in the json says accepted (without md5) - does that make sense ? I have the feeling I am too confusing right now :)

deleting the json works fine and the md5 error is gone but I do not have the database so I can not check if it actually finishes.

Hope that helps

**Comment by fu** [comment]
ah I see that it is actually part of the script :)
`        converted_result = search_result_intermediate.replace('accepted', 'rejected')`
My guess filtered result never finished correctly ...

**Comment by davidyang10** [comment]
Hi Christian,

I deleted the *_rejected.csv.u.json that corresponds to the “rejected.csv” file I was using as an input file to the upeptide_mapper engine. The rejected input file to upeptide_mapper was an output file of the filter_csv engine. Deleting the rejected json as you suggested allows the upeptide_mapper engine to run through. You originally suggested that the upeptide_mapper engine was failing when given the rejected input file because the corresponding rejected json had the run_status for the filter_csv engine still set to “scheduled”, which would suggest that the filter_csv run did not complete. It appears that using filter_csv with the parameter write_unfiltered_results always produces two input jsons, one for the accepted file and one for the rejected file. The input json for the rejected file is produced by copying the input json for the accepted file in the filter_csv wrappers. The “run_status” is set to complete in the accepted json because that json gets updated after the filter_csv engine runs. The “run_status” is set to “scheduled” still for the rejection json because that input json does not get updated. 

I am still unsure why the file ending with rejected.csv causes a failure when used as an input to the upeptide_mapper engine. Previously, I have used rejected files produced by filter_csv where the corresponding jsons still have the “run_status” for filter_csv set to “scheduled” as input files to the engine merge_csv. However, I did not run into the issue I am having now with upeptide_mapper where I need to delete the rejected json in order for the engine to run through.

**Comment by fu** [comment]
Ok, I think I got it and you are right. The rejected file never got its appropriate json. I added a functionality to the unode so such case can be fixed in the wrappers. Checkout branch hotfix/rejected_csv for a fix.

**Comment by davidyang10** [comment]
Hi Christian,
It looks like your fix solves the issue with filter_csv. Thanks for your help and it would be great if the fix could be added in the master branch.

**Comment by davidyang10** [comment]
Hi Christian,

The function you added fix_md5_and_file_in_json works fine. However when I tried to use the engine filter_csv on a file without writing the unfiltered results it caused an error. I believe this is because when the write_unfiltered_results parameter is set to False, the output_file_unfiltered in the filter_csv wrapper is set to None. A json corresponding to "None" (which does not exist) is then passed to the fix_md5_and_file_in_json which causes the error. The filter step worked once I modified the filter_csv wrapper so that the unfiltered output file json only goes into the fix_md5_and_file_in_json function if the unfiltered output file is not None.
